<a href="https://colab.research.google.com/github/EnayathShaik/AI_ML/blob/Practice/social_network_ads_prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd

In [8]:
df = pd.read_csv('/content/Social_Network_Ads.csv')
df.shape

(400, 5)

In [9]:
df.head()

User ID  Gender   Age  EstimatedSalary  Purchased
0  15624510    Male  19.0          19000.0          0
1  15810944    Male  35.0          20000.0          0
2  15668575  Female  26.0          43000.0          0
3  15603246  Female  27.0          57000.0          0
4  15804002    Male  19.0          76000.0          0

In [10]:
df.describe()

User ID         Age  EstimatedSalary   Purchased
count  4.000000e+02  400.000000       400.000000  400.000000
mean   1.569154e+07   37.655000     69742.500000    0.357500
std    7.165832e+04   10.482877     34096.960282    0.479864
min    1.556669e+07   18.000000     15000.000000    0.000000
25%    1.562676e+07   29.750000     43000.000000    0.000000
50%    1.569434e+07   37.000000     70000.000000    0.000000
75%    1.575036e+07   46.000000     88000.000000    1.000000
max    1.581524e+07   60.000000    150000.000000    1.000000

In [11]:
df.drop(['User ID'], inplace = True, axis=1 )
df.head()

Gender   Age  EstimatedSalary  Purchased
0    Male  19.0          19000.0          0
1    Male  35.0          20000.0          0
2  Female  26.0          43000.0          0
3  Female  27.0          57000.0          0
4    Male  19.0          76000.0          0

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Gender           400 non-null    object 
 1   Age              400 non-null    float64
 2   EstimatedSalary  400 non-null    float64
 3   Purchased        400 non-null    int64  
dtypes: float64(2), int64(1), object(1)
memory usage: 12.6+ KB


In [13]:
df.isnull().sum()

Gender             0
Age                0
EstimatedSalary    0
Purchased          0
dtype: int64

In [15]:
df['Gender'] = df['Gender'].map({'Male':1, 'Female':0})

In [16]:
df.head()

Gender   Age  EstimatedSalary  Purchased
0       1  19.0          19000.0          0
1       1  35.0          20000.0          0
2       0  26.0          43000.0          0
3       0  27.0          57000.0          0
4       1  19.0          76000.0          0

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Gender           400 non-null    int64  
 1   Age              400 non-null    float64
 2   EstimatedSalary  400 non-null    float64
 3   Purchased        400 non-null    int64  
dtypes: float64(2), int64(2)
memory usage: 12.6 KB


In [18]:
x = df.iloc[:,:-1]
y = df.iloc[:, -1]
print(x.shape)
print(y.shape)

(400, 3)
(400,)


In [19]:
x.head()

Gender   Age  EstimatedSalary
0       1  19.0          19000.0
1       1  35.0          20000.0
2       0  26.0          43000.0
3       0  27.0          57000.0
4       1  19.0          76000.0

In [20]:
y.head()

0    0
1    0
2    0
3    0
4    0
Name: Purchased, dtype: int64

In [21]:
from sklearn.model_selection import train_test_split

In [24]:
x_train,x_test, y_train, y_test = train_test_split(x,y, test_size= .2, random_state= 42, stratify=y) #stratify maintains the ratio of 0s, 1s in training and test data set
print (x_train.shape)
print (y_train.shape)
print(x_test.shape)
print(y_test.shape)

(320, 3)
(320,)
(80, 3)
(80,)


In [29]:
import tensorflow as tf
from tensorflow import keras
from keras.layers import Dense
from tensorflow.keras import Sequential

In [30]:
model = Sequential()
model.add(Dense(128, activation= 'relu', input_dim = 3))
model.add(Dense(128, activation = 'relu'))
model.add(Dense(1, activation = 'sigmoid'))
model.summary()

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense (Dense)                        │ (None, 128)                 │             512 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ (None, 128)                 │          16,512 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_2 (Dense)                      │ (None, 1)                   │             129 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 17,153 (67.00 KB)

 Trainable params: 17,153 (67.00 KB)

 Non-trainable params: 0 (0.00 B)

In [31]:
model.compile(optimizer= 'adam', loss='binary_crossentropy', metrics=['accuracy'])

In [32]:
model.fit(x_train, y_train , epochs= 100, validation_data= (x_test, y_test))

Epoch 1/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 2s 49ms/step - accuracy: 0.5624 - loss: 1587.0730 - val_accuracy: 0.6375 - val_loss: 849.3397
Epoch 2/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 17ms/step - accuracy: 0.6046 - loss: 984.3078 - val_accuracy: 0.3625 - val_loss: 76.3486
Epoch 3/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 18ms/step - accuracy: 0.5234 - loss: 331.8400 - val_accuracy: 0.3625 - val_loss: 7.0689
Epoch 4/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 14ms/step - accuracy: 0.5284 - loss: 311.2762 - val_accuracy: 0.3625 - val_loss: 195.1076
Epoch 5/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 18ms/step - accuracy: 0.5089 - loss: 103.1508 - val_accuracy: 0.3625 - val_loss: 84.7005
Epoch 6/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 18ms/step - accuracy: 0.4795 - loss: 64.1596 - val_accuracy: 0.3625 - val_loss: 124.8614
Epoch 7/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 17ms/step - accuracy: 0.4106 - loss: 177.5154 - val_accuracy: 0.3625 - val_loss: 455.4916
Epoch 8/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - accuracy: 0.4519 - loss: 3

In [33]:
# Applying feature scaling to improve accuracy

In [34]:
from sklearn.preprocessing import StandardScaler

In [35]:
std_scaler = StandardScaler()

In [37]:
x_train_sc = std_scaler.fit_transform(x_train)
x_test_sc = std_scaler.transform(x_test)

In [38]:
model_sc  = Sequential()
model_sc.add(Dense(128, activation= 'relu', input_dim = 3))
model_sc.add(Dense(128, activation = 'relu'))
model_sc.add(Dense(1, activation = 'sigmoid'))
model_sc.summary()

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense_3 (Dense)                      │ (None, 128)                 │             512 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_4 (Dense)                      │ (None, 128)                 │          16,512 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_5 (Dense)                      │ (None, 1)                   │             129 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 17,153 (67.00 KB)

 Trainable params: 17,153 (67.00 KB)

 Non-trainable params: 0 (0.00 B)

In [39]:
model_sc.compile(optimizer='adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

In [40]:
history_sc = model_sc.fit(x_train_sc, y_train, epochs=100, validation_data=(x_test_sc,y_test))

Epoch 1/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 3s 81ms/step - accuracy: 0.6801 - loss: 0.6491 - val_accuracy: 0.7625 - val_loss: 0.5644
Epoch 2/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 1s 23ms/step - accuracy: 0.8738 - loss: 0.4806 - val_accuracy: 0.8250 - val_loss: 0.4728
Epoch 3/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 14ms/step - accuracy: 0.8397 - loss: 0.4005 - val_accuracy: 0.8250 - val_loss: 0.4067
Epoch 4/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 15ms/step - accuracy: 0.8910 - loss: 0.3097 - val_accuracy: 0.8375 - val_loss: 0.3664
Epoch 5/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.8697 - loss: 0.3132 - val_accuracy: 0.8625 - val_loss: 0.3390
Epoch 6/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.8674 - loss: 0.2979 - val_accuracy: 0.8875 - val_loss: 0.3201
Epoch 7/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - accuracy: 0.8785 - loss: 0.2533 - val_accuracy: 0.9125 - val_loss: 0.3010
Epoch 8/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - accuracy: 0.8698 - loss: 0.2764 - val_accuracy: 0.88

**Without Scaling**

Epoch 100/100
10/10 - accuracy: 0.4819 - loss: 156.6003 - val_accuracy: 0.3625 - val_loss: 101.0254

**With Scaling**

Epoch 100/100
10/10 - accuracy: 0.9422 - loss: 0.1460 - val_accuracy: 0.9000 - val_loss: 0.2589

## Weights initialization

In [42]:
model_wts = Sequential()
model_wts.add(Dense(128, activation = 'relu', input_dim = 3, kernel_initializer='glorot_normal'))
model_wts.add(Dense(128, activation = 'relu', kernel_initializer='glorot_normal'))
model_wts.add(Dense(1, activation='sigmoid'))
model_wts.summary()

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense_9 (Dense)                      │ (None, 128)                 │             512 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_10 (Dense)                     │ (None, 128)                 │          16,512 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_11 (Dense)                     │ (None, 1)                   │             129 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 17,153 (67.00 KB)

 Trainable params: 17,153 (67.00 KB)

 Non-trainable params: 0 (0.00 B)

In [44]:
model_wts.compile(optimizer= 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])
model_wts.fit(x_train_sc, y_train, epochs= 100, validation_data = (x_test_sc, y_test))

Epoch 1/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 3s 47ms/step - accuracy: 0.6031 - loss: 0.6627 - val_accuracy: 0.7750 - val_loss: 0.5569
Epoch 2/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 18ms/step - accuracy: 0.8526 - loss: 0.5077 - val_accuracy: 0.8250 - val_loss: 0.4661
Epoch 3/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 17ms/step - accuracy: 0.8488 - loss: 0.4147 - val_accuracy: 0.8750 - val_loss: 0.3990
Epoch 4/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 17ms/step - accuracy: 0.8453 - loss: 0.3711 - val_accuracy: 0.8750 - val_loss: 0.3539
Epoch 5/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 17ms/step - accuracy: 0.8792 - loss: 0.3081 - val_accuracy: 0.9000 - val_loss: 0.3238
Epoch 6/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 15ms/step - accuracy: 0.8873 - loss: 0.2766 - val_accuracy: 0.9000 - val_loss: 0.3121
Epoch 7/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 17ms/step - accuracy: 0.8816 - loss: 0.2836 - val_accuracy: 0.8875 - val_loss: 0.2916
Epoch 8/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 18ms/step - accuracy: 0.8686 - loss: 0.2894 - val_accuracy: 0.